lib/exprotobuf.ex

```diff
@@ -60,52 +54,77 @@ defmodule Protobuf do
   defp parse(%Config{namespace: ns, schema: schema, inject: inject, from_file: nil}, _) do
     Parser.parse!(schema) |> namespace_types(ns, inject)
   end
-  defp parse(%Config{namespace: ns, schema: schema, inject: inject, from_file: file}, caller) do
-    {path, _} = Code.eval_quoted(file, [], caller)
-    path      = Path.expand(path) |> Path.dirname
-    opts      = [imports: [path]]
-    Parser.parse!(schema, opts) |> namespace_types(ns, inject)
+  defp parse(%Config{namespace: ns, inject: inject, from_file: file}, caller) do
+    {paths, import_dirs} = resolve_paths(file, caller)
+
+    Enum.reduce(paths, [], fn(path, defs) ->
+      schema = File.read!(path)
+      new_defs = Parser.parse!(schema, [imports: import_dirs]) |> namespace_types(ns, inject)
+      defs ++ new_defs
+    end)
   end
 
   # Apply namespace to top-level types
   defp namespace_types(parsed, ns, inject) do
+    prefix = namespace_prefix(parsed, ns, inject)
     for {{type, name}, fields} <- parsed do
-      if inject do
-        {{type, :"#{name |> normalize_name}"}, namespace_fields(type, fields, ns)}
-      else
-        {{type, :"#{ns}.#{name |> normalize_name}"}, namespace_fields(type, fields, ns)}
-      end
+      {{type, normalize_name(prefix, name)}, namespace_fields(type, fields, prefix)}
     end
   end
 
   # Apply namespace to nested types
-  defp namespace_fields(:msg, fields, ns), do: Enum.map(fields, &namespace_fields(&1, ns))
+  defp namespace_fields(:msg, fields, prefix), do: Enum.map(fields, &namespace_fields(&1, prefix))
   defp namespace_fields(_, fields, _),     do: fields
-  defp namespace_fields(field, ns) when not is_map(field) do
+  defp namespace_fields(field, prefix) when not is_map(field) do
     case elem(field, 0) do
-      :gpb_oneof -> field |> Utils.convert_from_record(OneOfField) |> namespace_fields(ns)
-      _          -> field |> Utils.convert_from_record(Field) |> namespace_fields(ns)
+      :gpb_oneof -> field |> Utils.convert_from_record(OneOfField) |> namespace_fields(prefix)
+      _          -> field |> Utils.convert_from_record(Field) |> namespace_fields(prefix)
     end
   end
-  defp namespace_fields(%Field{type: {type, name}} = field, ns) do
-    %{field | :type => {type, :"#{ns}.#{name |> normalize_name}"}}
+  defp namespace_fields(%Field{type: {type, name}} = field, prefix) do
+    %{field | :type => {type, normalize_name(prefix, name)}}
   end
-  defp namespace_fields(%Field{} = field, _ns) do
+  defp namespace_fields(%Field{} = field, _prefix) do
     field
   end
-  defp namespace_fields(%OneOfField{} = field, _ns) do
-    field |> Map.put(:fields, Enum.map(field.fields, &namespace_fields(&1, _ns)))
+  defp namespace_fields(%OneOfField{} = field, prefix) do
+    field |> Map.put(:fields, Enum.map(field.fields, &namespace_fields(&1, prefix)))
   end
 
+  defp namespace_prefix(parsed, _ns, true), do: ["Elixir"] ++ namespace_prefix(parsed)
```

Comment:
I think this can be written more efficiently as `["Elixir" | namespace_prefix(parsed)]`